File: managers/ModeManager.cs
using Vtt.Modes;

namespace Vtt.Managers;

public class ModeManager
{
    private static ModeManager? instance;

    private Mode? mode;

    private ModeManager()
    {
        mode = null;
    }

    public static ModeManager getInstance()
    {
        if (instance == null)
        {
            instance = new ModeManager();
        }
        return instance;
    }

    public Mode getCurrentMode()
    {
        if (mode == null)
        {
            throw new InvalidOperationException("Mode is not set!");
        }
        return mode;
    }

    public void setMode(Mode mode)
    {
        this.mode = mode;
    }
}

File: src/managers/mode_manager.rs
use crate::modes::Mode;

pub struct ModeManager {
    pub current_mode: Box<dyn Mode>
}

impl ModeManager {
    pub fn new(init_mode: Box<dyn Mode>) -> ModeManager { ModeManager {current_mode: init_mode} }
    pub fn set_mode(&mut self, new_mode: Box<dyn Mode>) {
        self.current_mode = new_mode;
    }
    pub fn get_current_mode(&mut self) -> &mut dyn Mode {
        self.current_mode.as_mut()
    }
}
